File: pathSum2.py
# problem: https://leetcode.com/problems/path-sum-ii/
# Runtime: 40 ms, faster than 88.81% of Python3 online submissions for Path Sum II.
# Memory Usage: 15.3 MB, less than 87.29% of Python3 online submissions for Path Sum II.

from typing import List

# Definition for a binary tree node.


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def traverse(self, node, total, stack):
        if not node:
            return None
        total += node.val
        stack.append(node.val)
        self.traverse(node.right, total, stack)
        self.traverse(node.left, total, stack)

        if not node.right and not node.left and total == self.sum:
            self.answer.append(stack.copy())
        stack.pop()
        total -= node.val

    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        self.answer = []
        self.sum = sum
        self.traverse(root, 0, [])

        return self.answer

File: pathSum2.cpp
// problem: https://leetcode.com/problems/path-sum-ii/
// Runtime: 15 ms, faster than 72.54% of C++ online submissions for Path Sum II.
// Memory Usage: 20.2 MB, less than 47.44% of C++ online submissions for Path
// Sum II.

#include <bits/stdc++.h>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
public:
  void dfs(TreeNode *root, vector<int> &stack, vector<vector<int>> &ans,
           int target) {
    if (!root)
      return;
    stack.push_back(root->val);
    if (!(root->left) && !(root->right) && target == root->val) {
      vector<int> tmp = stack;
      ans.push_back(tmp);
    }

    dfs(root->right, stack, ans, target - root->val);
    dfs(root->left, stack, ans, target - root->val);

    stack.pop_back();
  }

  vector<vector<int>> pathSum(TreeNode *root, int targetSum) {
    vector<vector<int>> ans;
    vector<int> stack;
    dfs(root, stack, ans, targetSum);
    return ans;
  }
};
